driver-core/src/main/java/com/datastax/driver/core/policies/RackAwarePolicy.java

```diff
@@ -0,0 +1,330 @@
+
+    final int startIdx = index.getAndIncrement();
+
+    return new AbstractIterator<Host>() {
+      private int idx = startIdx;
+      private int remainingLocal = localRackHosts.size();
+      private int remainingRemote = localDcRemoteHosts.size();
+      private Iterator<String> foreignDcs;
+      private List<Host> currentDcHosts;
+      private int currentDcRemaining;
+
+      @Override
+      protected Host computeNext() {
+        while (true) {
+          // Returns local dc and local rack hosts
+          if (remainingLocal > 0) {
+            remainingLocal--;
+            int c = idx++ % localRackHosts.size();
+            if (c < 0) {
+              c += localRackHosts.size();
+            }
+            return localRackHosts.get(c);
+          }
+
+          // Returns local dc and non-local rack hosts
+          if (remainingRemote > 0) {
+            remainingRemote--;
+            int c = idx++ % localDcRemoteHosts.size();
+            if (c < 0) {
+              c += localDcRemoteHosts.size();
+            }
+            return localDcRemoteHosts.get(c);
+          }
+
+          // Returns remaining hosts in remote datacenters
+          if (currentDcHosts != null && currentDcRemaining > 0) {
+            currentDcRemaining--;
+            int c = idx++ % currentDcHosts.size();
+            if (c < 0) {
+              c += currentDcHosts.size();
+            }
+            return currentDcHosts.get(c);
+          }
+
+          if (foreignDcs == null) {
+            Set<String> copy = new HashSet<String>(perDcLiveHosts.keySet());
+            copy.remove(localDc);
+            foreignDcs = copy.iterator();
+          }
+
+          if (!foreignDcs.hasNext()) break;
+
+          String nextRemoteDc = foreignDcs.next();
+          CopyOnWriteArrayList<Host> nextDcHosts = perDcLiveHosts.get(nextRemoteDc);
+          if (nextDcHosts != null) {
+            // Clone for thread safety
+            List<Host> dcHosts = cloneList(nextDcHosts);
+            currentDcHosts = dcHosts.subList(0, Math.min(dcHosts.size(), usedHostsPerRemoteDc));
+            currentDcRemaining = currentDcHosts.size();
+          }
+        }
+        return endOfData();
+      }
+    };
+  }
+
+  @Override
+  public void onAdd(Host host) {
+    onUp(host);
+  }
+
+  @Override
+  public void onUp(Host host) {
+    String dc = dc(host);
+    String rack = dc + "-" + rack(host);
+
+    if (localDc.equals(UNSET) && !dc.equals(UNSET)) {
+      localDc = dc;
+    }
+
+    if (localRack.equals(UNSET) && !rack(host).equals(UNSET)) {
+      logger.info(
+          "Using provided datacenter name '{}' and rack name '{}' for RackAwarePolicy (if this is incorrect, please provide the correct names with RackAwarePolicy constructor)",
+          dc,
+          rack(host));
+      localRack = rack;
+    }
+
+    addToConcurrentMapValues(perDcLiveHosts, dc, host);
+    addToConcurrentMapValues(perRackLiveHosts, rack, host);
+  }
```

Comment:
I could not find relevant documentation, but `onUp` and `onDown` functions are only called inside a `Manager` object which is responsible for fetching topology updates. The functions are called in case the manager receives an update event from the server, so I guess practically this case should not happen.